### PR TITLE
feat: robust confidence sequence estimator

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  double k;
+  double log_k;
+  double sum_x;
+  double sum_low_v;
+  double sum_mid_v;
+  std::map<std::pair<uint64_t, bool>, double> sum_v_histo;
+  uint64_t t;
+};
+
+class countable_discrete_base
```
Comment (lalo) [inline vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:37]: what is the reason behind this name?

Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:37]: this is based on some of Paul's naming conventions

File: .vscode/settings.json [hunk 2]
```diff
@@ -6,5 +6,88 @@
             ],
             "url": "./test/vwtest.schema.json"
         }
-    ]
+    ],
```
Comment (lalo) [inline .vscode/settings.json:9]: added by mistake probs

Comment (bassmang) [reply .vscode/settings.json]: yup revered

File: vowpalwabbit/core/src/confidence_sequence_robust.cc [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+    double s_0, double thres, std::map<uint64_t, double>& memo, double a, double b, double toll_x, double toll_f) const
+{
+  auto f = [this, &s_0, &thres, &memo](double mu) -> double { return log_wealth_mix(mu, s_0, thres, memo) - thres; };
+  double fa = f(a);
+  double fb = f(b);
+  double fs;
+
+  if (!(fa * fb < 0))
+  {
+    // std::cerr << "Signs of f(x_min) and f(x_max) must be opposites" << std::endl;
```
Comment (lalo) [inline vowpalwabbit/core/src/confidence_sequence_robust.cc:181]: do we still need this if?

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h (hunk 1 above)
Comment (lalo) [reply vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:40]: maybe add a comment referencing on top of this class or whichever is more relevant to that other repo with the robust impl of Paul

File: vowpalwabbit/core/src/confidence_sequence_robust.cc [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+    if (upper_bound < thres) { return upper_bound; }
+
+    double v = get_v_impl(memo, j);
+    log_es.push_back(get_lam_sqrt_tp1(j) * y - v);
+    log_ws.push_back(get_log_weight(j));
+    j += 1;
+  }
+}
+
+double countable_discrete_base::root_brentq(
```
Comment (lalo) [inline vowpalwabbit/core/src/confidence_sequence_robust.cc:171]: adding a unit test for root_brentq might be a good idea, just like the simple one you had and shared before

File: vowpalwabbit/core/src/confidence_sequence_robust.cc [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+  }
+  return ret_val;
+}
+
+void g_tilde::histo_insert(double x)
+{
+  uint64_t n = std::floor(std::log(x) / log_k);
+  double x1 = std::pow(k, n);
+  double alpha = (k * x1 - x) / ((k - 1.0) * x1);
+  // TODO: Kahan summation
```
Comment (lalo) [inline vowpalwabbit/core/src/confidence_sequence_robust.cc:43]: is this comment stale?

File: vowpalwabbit/core/src/confidence_sequence_robust.cc (hunk 3 above)
Comment (bassmang) [reply vowpalwabbit/core/src/confidence_sequence_robust.cc]: Yes that check should be there, uncommented

File: vowpalwabbit/core/src/confidence_sequence_robust.cc (hunk 5 above)
Comment (bassmang) [reply vowpalwabbit/core/src/confidence_sequence_robust.cc:43]: It's copied from Paul's Python code, not sure if its still needed

File: vowpalwabbit/core/src/confidence_sequence_robust.cc (hunk 4 above)
Comment (bassmang) [reply vowpalwabbit/core/src/confidence_sequence_robust.cc:171]: added check

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h (hunk 1 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:43]: added comments connecting the code

**Review by lalo — APPROVED** [review]

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h [hunk 6]
```diff
@@ -0,0 +1,103 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#pragma once
+
+#if !defined(__APPLE__) && !defined(_WIN32)
+#  define __STDCPP_MATH_SPEC_FUNCS__ 201003L
+#  define __STDCPP_WANT_MATH_SPEC_FUNCS__ 1
+#endif
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:10]: Please confine this to just the cc

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+#include "vw/core/io_buf.h"
+#include "vw/core/metric_sink.h"
+
+constexpr double CS_ROBUST_DEFAULT_ALPHA = 0.05f;
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:15]: Put this in a details namespace

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  void reset_stats();
+
+  const double k;
+  const double log_k;
+
+  double sum_x;
+  double sum_low_v;
+  double sum_mid_v;
+  uint64_t t;
+  std::map<std::pair<uint64_t, bool>, double> sum_v_histo;
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:39]: Missing map and utility includes

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+#pragma once
+
+#if !defined(__APPLE__) && !defined(_WIN32)
+#  define __STDCPP_MATH_SPEC_FUNCS__ 201003L
+#  define __STDCPP_WANT_MATH_SPEC_FUNCS__ 1
+#endif
+
+#include "vw/core/io_buf.h"
+#include "vw/core/metric_sink.h"
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:13]: These should be replaced with vw_fwd include

File: test/unit_test/confidence_sequence_robust_test.cc [hunk 10]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/confidence_sequence_robust.h"
```
Comment (jackgerrits) [inline test/unit_test/confidence_sequence_robust_test.cc:5]: New unit tests should go in `vowpalwabbit/core/tests`

**Review by jackgerrits — CHANGES_REQUESTED** [review]

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  g_tilde(double k);
+  double histo_variance(double lam_sqrt_tp1) const;
+  void histo_insert(double x);
+  void add_obs(double x);
+  double get_s() const;
+  double get_v(double lam_sqrt_tp1) const;
+  void reset_stats();
+
+  const double k;
+  const double log_k;
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:33]: Const means this class cannot be moved. Is that what is indended?

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h (hunk 6 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:10]: moved

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h (hunk 8 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:39]: added

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h (hunk 11 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:33]: I think the const needs to be removed

File: vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+#include "vw/core/io_buf.h"
+#include "vw/core/metric_sink.h"
+
+constexpr double CS_ROBUST_DEFAULT_ALPHA = 0.05f;
+
+namespace VW
+{
+
+// Equivalent to GTilde in confidence_sequence_robust.py
+class g_tilde
```
[the remark this answers is not in the document]
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/confidence_sequence_robust.h:21]: sure, g_tilde and countable_discrete_base can be moved to details

**Review by jackgerrits — APPROVED** [review]

**Pull request merged**